core/src/main/scala/chisel3/Disable.scala

```diff
@@ -0,0 +1,77 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package chisel3
+
+import chisel3.internal._
+import chisel3.experimental.{IntrinsicModule, OpaqueType, SourceInfo}
+import chisel3.internal.sourceinfo.SourceInfoTransform
+
+import scala.language.experimental.macros
+import scala.collection.immutable.ListMap
+
+/** API for handling disabling of simulation constructs
+  *
+  * Disables may be non-synthesizable so they can only be used for disabling simulation constructs
+  *
+  * The default disable is the "hasBeenReset" of the currently in scope reset.
+  * It can be set by the user via the [[withDisable]] API
+  *
+  * Users can access the current `Disable` with [[Module.disable]]
+  */
+// We could just an OpaqueType, but since OpaqueTypes have some API holes, this non-Data type
+class Disable private[chisel3] (private[chisel3] val value: Bool) {
+
+  /** Logical not
+    *
+    * @return invert the logical value of this `Disable`
+    * @group Bitwise
+    */
+  final def unary_! : Disable = macro SourceInfoTransform.noArg
+
+  /** @group SourceInfoTransformMacro */
+  def do_unary_!(implicit sourceInfo: SourceInfo): Disable = new Disable(!this.value)
+}
+
+object Disable {
+
+  sealed trait Option
+  case object None extends Option
+  case object NotHasBeenReset extends Option
```

Comment:
The shadowing of `Option` isn't awesome. I'll usually use `sealed trait Type` for this pattern.

Reply:
Yeah I was not sure how I felt about the names, I'll use `Type` instead of `Option`. Do you have an idea for better name than `None` since that shadowing isn't great either? I thought about `No`, could also do `NoDisable` but the normal use of this API will be `Disable.NoDisable` so that might be a bit much (but is also maybe okay).

Reply:
You could use `Type` as a sealed trait and as an object. You then get `Type` as a package-like separator to avoid confusing around `Disable.NoDisable`. Just an idea:

```scala
sealed trait Type
object Type {
  case object NoDisable extends Type
  case object NotHasBeenReset extends Type
}
```

`NoDisable` seems okay. Also, `None` isn't quite as bad if wrapped in the `Type` object.

Reply:
Using `Type`, `Never`, and `BeforeReset` (great suggestion on that one!)